source/particle/interpolator/quadratic_least_squares.cc

```diff
@@ -0,0 +1,302 @@
+          particle_handler.particles_in_cell(found_cell);
+
+
+        std::vector<std::vector<double> > cell_properties(positions.size(),
+                                                          std::vector<double>(n_particle_properties,
+                                                                              numbers::signaling_nan<double>()));
+
+        const unsigned int n_particles = std::distance(particle_range.begin(), particle_range.end());
+
+        const unsigned int n_matrix_columns = (dim == 2) ? 6 : 10;
+        AssertThrow(n_particles >= n_matrix_columns,
+                    ExcMessage("At least one cell didn't contain enough particles to interpolate a unique solution for the cell."
+                               " The 'quadratic' interpolation scheme does not support this case."));
+
+
+        // Notice that the size of matrix A is n_particles x n_matrix_columns
+        // which usually is not a square matrix. Therefore, we find the
+        // least squares solution of Ac=r by solving the reduced QR factorization
+        // Ac = QRc = b -> Q^TQRc = Rc =Q^Tb
+        dealii::ImplicitQR<dealii::Vector<double>> qr;
+        // A is a std::vector of Vectors(which are it's columns) so that we create what the ImplicitQR
+        // class needs.
+        std::vector<dealii::Vector<double>> A(n_matrix_columns, dealii::Vector<double>(n_particles));
+        std::vector<Vector<double>> b(n_particle_properties, Vector<double>(n_particles));
+        std::vector<Vector<double>> QTb(n_particle_properties, Vector<double>(n_matrix_columns));
+        std::vector<Vector<double>> c(n_particle_properties, Vector<double>(n_matrix_columns));
+        for (unsigned int property_index = 0; property_index < n_particle_properties; ++property_index)
+          if (selected_properties[property_index])
+            b[property_index] = 0;
+
+        unsigned int positions_index = 0;
+        const double cell_diameter = found_cell->diameter();
+        for (typename ParticleHandler<dim>::particle_iterator particle = particle_range.begin();
+             particle != particle_range.end(); ++particle, ++positions_index)
+          {
+            const auto &particle_property_value = particle->get_properties();
+            for (unsigned int property_index = 0; property_index < n_particle_properties; ++property_index)
+              if (selected_properties[property_index])
+                b[property_index][positions_index] = particle_property_value[property_index];
+            const Tensor<1, dim, double> relative_particle_position = (particle->get_location() - approximated_cell_midpoint) / cell_diameter;
+            // A is accessed by A[column][row] here since we need to append
+            // columns into the qr matrix.
+
+            // There is a potential that in the future we may need to add more
+            // 'mixed' terms to the interpolation for some problems. We for now
```

Comment:
I don't get this comment. What terms are you thinking of? You've covered all quadratic mixed terms...

Reply:
This comment was meant to answer @gassmoeller 's suggestion to include in 2D (x^2)y x(y^2) and (xy)^2 terms. The corresponding change to 3D would cause it to have 27 total terms(all combinations of a,b,c in {0,1,2} generating (x^a)(y^b)(z^c)). @egpuckett and I haven't found a case so far that has needed these terms (or a number of particles per cell that increases with mesh refinement) to get the expected convergence, but after Rene's comment I wanted to acknowledge that there may be a case where those terms would provide additional accuracy and after finding such case we would want to add those terms. The drawback of adding those terms is the decreased efficiency of using more particles per cell.

Reply:
Oh, I see -- you're projecting onto the space $P_2$, not $Q_2$. I guess that's legitimate.

Can you incorporate a version of your answer here into the comment in question?